core/executor/src/wasm_executor.rs

```diff
@@ -124,1201 +270,1066 @@ fn deadline_to_timestamp(deadline: u64) -> Option<offchain::Timestamp> {
-	},
-	// return 0 and place u32::max_value() into written_out if no value exists for the key.
-	ext_get_allocated_child_storage(
-		storage_key_data: *const u8,
-		storage_key_len: u32,
-		key_data: *const u8,
-		key_len: u32,
-		written_out: *mut u32
-	) -> *mut u8 => {
-		let storage_key = this.memory.get(
-			storage_key_data,
-			storage_key_len as usize
-		).map_err(|_| "Invalid attempt to determine storage_key in ext_get_allocated_child_storage")?;
-		let key = this.memory.get(
-			key_data,
-			key_len as usize
-		).map_err(|_| "Invalid attempt to determine key in ext_get_allocated_child_storage")?;
-
-		let maybe_value = {
-			let storage_key = ChildStorageKey::from_slice(&storage_key)
-				.ok_or_else(|| "ext_get_allocated_child_storage: child storage key is not valid")?;
-			this.ext.child_storage(storage_key, &key)
-		};
 
-		debug_trace!(
-			target: "wasm-trace", "*** Getting child storage: {} -> {} == {}   [k={}]",
-			primitives::hexdisplay::ascii_format(&storage_key),
-			if let Some(_preimage) = this.hash_lookup.get(&key) {
-				format!("%{}", ::primitives::hexdisplay::ascii_format(&_preimage))
+		ext_storage_changes_root(
+			parent_hash_data: Pointer<u8>,
+			_len: WordSize,
+			result: Pointer<u8>,
+		) -> u32 {
+			let mut parent_hash = [0u8; 32];
+			context.read_memory_into(parent_hash_data, &mut parent_hash[..])
+				.map_err(|_| "Invalid attempt to get parent_hash in ext_storage_changes_root")?;
+			let r = runtime_io::storage_changes_root(parent_hash);
+
+			if let Some(r) = r {
+				context.write_memory(result, &r[..])
+					.map_err(|_| "Invalid attempt to set memory in ext_storage_changes_root")?;
+				Ok(1)
 			} else {
-				format!(" {}", ::primitives::hexdisplay::ascii_format(&key))
-			},
-			if let Some(ref b) = maybe_value {
-				&format!("{}", HexDisplay::from(b))
+				Ok(0)
+			}
+		}
+
+		ext_blake2_256_enumerated_trie_root(
+			values_data: Pointer<u8>,
+			lens_data: Pointer<u32>,
+			lens_len: WordSize,
+			result: Pointer<u8>,
+		) {
+			let values = (0..lens_len)
+				.map(|i| context.read_primitive(lens_data.offset(i).ok_or("Pointer overflow")?))
+				.collect::<std::result::Result<Vec<u32>, _>>()?
+				.into_iter()
+				.scan(0u32, |acc, v| { let o = *acc; *acc += v; Some((o, v)) })
+				.map(|(offset, len)|
+					context.read_memory(values_data.offset(offset).ok_or("Pointer overflow")?, len)
+						.map_err(|_|
+							"Invalid attempt to get memory in ext_blake2_256_enumerated_trie_root"
+						)
+				)
+				.collect::<std::result::Result<Vec<_>, _>>()?;
+			let r = Layout::<Blake2Hasher>::ordered_trie_root(values.into_iter());
+			context.write_memory(result, &r[..])
+				.map_err(|_| "Invalid attempt to set memory in ext_blake2_256_enumerated_trie_root")?;
+			Ok(())
+		}
+
+		ext_chain_id() -> u64 {
+			Ok(runtime_io::chain_id())
+		}
+
+		ext_twox_64(data: Pointer<u8>, len: WordSize, out: Pointer<u8>) {
+			let result: [u8; 8] = if len == 0 {
+				let hashed = twox_64(&[0u8; 0]);
+				hashed
```

Comment:
So this effectively removes caching logic. Did you find out why it was there and whether we can remove it? If so, I'd be curious to hear

Reply:
I did not understand why this logic is required? The native api also does not have any caching logic :woman_shrugging: 

When we want to move to generating all this interface stuff, we can not have all these edge cases anyway.

Reply:
> I did not understand why this logic is required? The native api also does not have any caching logic

Yeah, I am confused by this as well and that is the reason why I am curious. I also cannot come up why it was there in the first place.

And at the same time, I am a bit of unsure that it was there for nothing. If you want I can try to find it out myself

Reply:
Do you mean the `hash_lookup` stuff? I think that was *just* for `debug_trace` readability. I agree with removing it.

Reply:
I just did some git archeology, and it seems that `hash_lookup` was introduced [for debugging](https://github.com/paritytech/substrate/commit/15f34526bc31161da6f27ef2a03fb00482993f86). I am good with removing it

Reply:
Yeah, as @jimpo highlighted it was mostly used in debug traces. All tests are passing, so I really hope nobody added this without any test.